util/codec/codec.go

```diff
@@ -284,105 +265,61 @@ func sizeInt(comparable bool) int {
-			if !mysql.HasUnsignedFlag(allTypes[i].Flag) {
-				b = encodeSignedInt(b, row.GetInt64(i), comparable)
-				break
-			}
-			// encode unsigned integers.
-			integer := row.GetInt64(i)
-			if integer < 0 {
-				b = encodeUnsignedInt(b, uint64(integer), comparable)
-			} else {
-				b = encodeSignedInt(b, integer, comparable)
-			}
-		case mysql.TypeFloat:
-			b = append(b, floatFlag)
-			b = EncodeFloat(b, float64(row.GetFloat32(i)))
-		case mysql.TypeDouble:
-			b = append(b, floatFlag)
-			b = EncodeFloat(b, row.GetFloat64(i))
-		case mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-			b = encodeBytes(b, row.GetBytes(i), comparable)
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
-			b = append(b, uintFlag)
-			t := row.GetTime(i)
-			// Encoding timestamp need to consider timezone.
-			// If it's not in UTC, transform to UTC first.
-			if t.Type == mysql.TypeTimestamp && sc.TimeZone != time.UTC {
-				err = t.ConvertTimeZone(sc.TimeZone, time.UTC)
-				if err != nil {
-					return nil, errors.Trace(err)
+			flag := varintFlag
+			if mysql.HasUnsignedFlag(allTypes[i].Flag) {
+				if integer := row.GetInt64(i); integer < 0 {
+					flag = uvarintFlag
 				}
 			}
-			var v uint64
-			v, err = t.ToPackedUint()
+			_, err = w.Write([]byte{flag})
 			if err != nil {
-				return nil, errors.Trace(err)
+				return errors.Trace(err)
 			}
-			b = EncodeUint(b, v)
-		case mysql.TypeDuration:
-			// duration may have negative value, so we cannot use String to encode directly.
-			b = append(b, durationFlag)
-			b = EncodeInt(b, int64(row.GetDuration(i, 0).Duration))
+			_, err = row.WriteTo(i, w)
 		case mysql.TypeNewDecimal:
-			b = append(b, decimalFlag)
 			// If hash is true, we only consider the original value of this decimal and ignore it's precision.
 			dec := row.GetMyDecimal(i)
-			bin, err := dec.ToHashKey()
+			var bin []byte
+			bin, err = dec.ToHashKey()
 			if err != nil {
-				return nil, errors.Trace(err)
+				return errors.Trace(err)
 			}
-			b = append(b, bin...)
-		case mysql.TypeEnum:
-			b = encodeUnsignedInt(b, uint64(row.GetEnum(i).ToNumber()), comparable)
-		case mysql.TypeSet:
-			b = encodeUnsignedInt(b, uint64(row.GetSet(i).ToNumber()), comparable)
-		case mysql.TypeBit:
-			// We don't need to handle errors here since the literal is ensured to be able to store in uint64 in convertToMysqlBit.
-			var val uint64
-			val, err = types.BinaryLiteral(row.GetBytes(i)).ToInt(sc)
-			terror.Log(errors.Trace(err))
-			b = encodeUnsignedInt(b, val, comparable)
-		case mysql.TypeJSON:
-			b = append(b, jsonFlag)
-			j := row.GetJSON(i)
-			b = append(b, j.TypeCode)
-			b = append(b, j.Value...)
+			_, err = w.Write(bin)
 		default:
-			return nil, errors.Errorf("unsupport column type for encode %d", allTypes[i].Tp)
+			_, err = row.WriteTo(i, w)
```

Comment:
It's dangerous to skip the type interpretation and hash the raw data directly.
For example:
```
create table t1(x float);
create table t2(x double);
insert into t1 values (1);
insert into t2 values (1);
select * from t1, t2 where t1.x=t2.x;
```

Reply:
Since that, I've no good solution except changing all `encodeXXX` function along with an `io.Writer` argument.
I'll revert this part code for now and try to optimize it in further PR.
Any suggestion is very welcome, @qw4990 @zz-jason @XuHuaiyu